node/network/approval-distribution/src/lib.rs

```diff
@@ -984,228 +1337,332 @@ impl State {
 				}
+
+				block = entry.parent_hash.clone();
 			}
 		}
 
-		if !assignments.is_empty() {
-			gum::trace!(
-				target: LOG_TARGET,
-				num = assignments.len(),
-				?num_blocks,
-				?peer_id,
-				"Sending assignments to a peer",
-			);
-
+		if !assignments_to_send.is_empty() {
 			ctx.send_message(NetworkBridgeMessage::SendValidationMessage(
 				vec![peer_id.clone()],
 				protocol_v1::ValidationProtocol::ApprovalDistribution(
-					protocol_v1::ApprovalDistributionMessage::Assignments(assignments),
+					protocol_v1::ApprovalDistributionMessage::Assignments(assignments_to_send),
 				),
 			))
 			.await;
 		}
 
-		if !approvals.is_empty() {
-			gum::trace!(
-				target: LOG_TARGET,
-				num = approvals.len(),
-				?num_blocks,
-				?peer_id,
-				"Sending approvals to a peer",
-			);
-
+		if !approvals_to_send.is_empty() {
 			ctx.send_message(NetworkBridgeMessage::SendValidationMessage(
-				vec![peer_id],
+				vec![peer_id.clone()],
 				protocol_v1::ValidationProtocol::ApprovalDistribution(
-					protocol_v1::ApprovalDistributionMessage::Approvals(approvals),
+					protocol_v1::ApprovalDistributionMessage::Approvals(approvals_to_send),
 				),
 			))
 			.await;
 		}
 	}
+
+	async fn enable_aggression(
+		&mut self,
+		ctx: &mut (impl SubsystemContext<Message = ApprovalDistributionMessage>
+		          + overseer::SubsystemContext<Message = ApprovalDistributionMessage>),
+		do_resend: bool,
+		metrics: &Metrics,
+	) {
+		let min_age = self.blocks_by_number.iter().next().map(|(num, _)| num);
+		let max_age = self.blocks_by_number.iter().rev().next().map(|(num, _)| num);
+		let config = self.aggression_config.clone();
+
+		let (min_age, max_age) = match (min_age, max_age) {
+			(Some(min), Some(max)) => (min, max),
+			_ => return, // empty.
+		};
+
+		let diff = max_age - min_age;
+		if !self.aggression_config.is_age_relevant(diff) {
+			return
+		}
+
+		adjust_required_routing_and_propagate(
+			ctx,
+			&mut self.blocks,
+			&self.topologies,
+			|block_entry| {
+				let block_age = max_age - block_entry.number;
+
+				if do_resend &&
+					config
+						.resend_unfinalized_period
+						.as_ref()
+						.map_or(false, |p| block_age > 0 && block_age % p == 0)
+				{
+					// Retry sending to all peers.
+					for (_, knowledge) in block_entry.known_by.iter_mut() {
+						knowledge.sent = Default::default();
+					}
+
+					return true
+				}
+
+				// Ramp up aggression only for the very oldest block(s).
+				// Approval voting can get stuck on a single block preventing
+				// its descendants from being finalized. Waste minimal bandwidth
+				// this way. Also, disputes might prevent finality - again, nothing
+				// to waste bandwidth on newer blocks for.
+				&block_entry.number == min_age
+			},
+			|required_routing, local, _| {
+				// It's a bit surprising not to have a topology at this age.
+				if *required_routing == RequiredRouting::PendingTopology {
+					gum::debug!(
+						target: LOG_TARGET,
+						age = ?diff,
```

Comment:
```suggestion
						age = ?diff,
						?relay_parent = block_entry.parent_hash,
```

Reply:
Block entry isn't in scope here, unfortunately.